src/language/server.ts

```diff
@@ -277,13 +282,61 @@ connection.onCompletionResolve(
 /**
  * Execute the entire playground script.
  */
-connection.onRequest('executeAll', async (params) => {
-  const connectionOptions = params.connectionOptions || {};
-  const runtime = new ElectronRuntime(
-    await CliServiceProvider.connect(params.connectionString, connectionOptions)
-  );
+connection.onRequest('executeAll', (params) => {
+  // Use Node worker threads to isolate each run of a playground
+  // to be able to cancel evaluation of infinite loops
+  const worker = new WorkerThreads(path.resolve(__dirname, 'worker.js'), {
+    workerData: {
+      codeToEvaluate: params.codeToEvaluate,
+      connectionString: params.connectionString,
+      connectionOptions: params.connectionOptions
+    }
+  });
+
+  // Listen for results from the worker thread
+  worker.on('message', (response) => {
+    // Print a debug message to the language server output
+    if (response.message) {
+      connection.console.log(`${response.message}`);
+    }
 
-  return await runtime.evaluate(params.codeToEvaluate);
+    // Send error message to the language server client
+    if (response.error) {
+      connection.sendNotification(
+        'showErrorMessage',
+        response.error.message
+      );
+    }
+
+    // Send result to the language server client
+    connection.sendNotification(
+      'executeAllDone',
+      response.result
+    );
+  });
+
+  // Listen for cancellation request from the language server client
+  connection.onRequest('cancelAll', async () => {
+    connection.console.log('Playground cancellation requested');
+    connection.sendNotification(
+      'showInformationMessage',
+      'The long running playground operation was canceled'
+    );
+
+    // Try to close mongoClient to free resources
+    worker.postMessage('terminate');
+
+    // Closing mongoClient...
+    // We can't wait for the actual result from cancelAll() function
+    // because it might never return a result in case of infinite loops
+    await sleep(3000);
```

Comment:
I have found a way to call `mongoClient.close` before the termination of the worker. This delay is not visible for a user since the cancelation message already sent to UI. We need this delay to let mongoClient to be closed before we terminate the worker. I am open to suggestions.

Reply:
I don't see anything bad with that, except that probably we don't need it.

I'd expect `terminate()` to close everything, socket connections included. If that is the case the client would be terminated, just not gracefully.

In case you'd like to close the connection more gently that seems ok to me.

Reply:
If you find this message passing with events a bit messy there are few libraries that helps a bit: https://www.npmjs.com/package/comlink

Reply:
I prefer using native modules when possible to not add any shady and unexpected behavior to the project. And to be honest I took a look at this package and it looks more confusing to me than original API 😀